src/mito2/src/manifest/checkpointer.rs

```diff
@@ -0,0 +1,150 @@
+#[derive(Debug)]
+pub(crate) struct Checkpointer {
+    region_id: RegionId,
+    manifest_options: RegionManifestOptions,
+    manifest_store: Arc<ManifestObjectStore>,
+    last_checkpoint_version: Arc<AtomicU64>,
+    is_doing_checkpoint: Arc<AtomicBool>,
+}
+
+impl Checkpointer {
+    pub(crate) fn new(
+        region_id: RegionId,
+        manifest_options: RegionManifestOptions,
+        manifest_store: Arc<ManifestObjectStore>,
+        last_checkpoint_version: ManifestVersion,
+    ) -> Self {
+        Self {
+            region_id,
+            manifest_options,
+            manifest_store,
+            last_checkpoint_version: Arc::new(AtomicU64::new(last_checkpoint_version)),
+            is_doing_checkpoint: Arc::new(AtomicBool::new(false)),
+        }
+    }
+
+    pub(crate) fn last_checkpoint_version(&self) -> ManifestVersion {
+        self.last_checkpoint_version.load(Ordering::Relaxed)
+    }
+
+    /// Check if it's needed to do checkpoint for the region by the checkpoint distance.
+    /// If needed, and there's no currently running checkpoint task, it will start a new checkpoint
+    /// task running in the background.
+    pub(crate) fn maybe_do_checkpoint(&self, manifest: &RegionManifest) {
+        if self.manifest_options.checkpoint_distance == 0 {
+            return;
+        }
+
+        let last_checkpoint_version = self.last_checkpoint_version();
+        if manifest.manifest_version - last_checkpoint_version
+            < self.manifest_options.checkpoint_distance
+        {
+            return;
+        }
+
+        // We can simply check whether there's a running checkpoint task like this, all because of
+        // the caller of this function is ran single threaded: it's in the region's worker loop.
+        if self.is_doing_checkpoint.load(Ordering::Relaxed) {
+            return;
+        }
+
+        let start_version = if last_checkpoint_version == 0 {
+            // Checkpoint version can't be zero by implementation.
+            // So last checkpoint version is zero means no last checkpoint.
+            MIN_VERSION
+        } else {
+            last_checkpoint_version + 1
+        };
+        let end_version = manifest.manifest_version;
+        info!(
+            "Start doing checkpoint for region {}, compacted version: [{}, {}]",
+            self.region_id, start_version, end_version,
+        );
+
+        let checkpoint = RegionCheckpoint {
+            last_version: end_version,
+            compacted_actions: (end_version - start_version + 1) as usize,
+            checkpoint: Some(manifest.clone()),
+        };
+        self.do_checkpoint(checkpoint);
+    }
+
+    fn do_checkpoint(&self, checkpoint: RegionCheckpoint) {
+        let is_doing_checkpoint = self.is_doing_checkpoint.clone();
+        is_doing_checkpoint.store(true, Ordering::Relaxed);
+        let guard = scopeguard::guard(is_doing_checkpoint, |x| {
+            x.store(false, Ordering::Relaxed);
+        });
```

Comment:
The guard is quite easy to implement. Do we need to introduce a new crate dependency for this? cc @tisonkun

Reply:
We already `scopeguard` in our deps, search it in the cargo.lock and you'll see. Besides, `scopeguard` is quite popular, it's top 100 crates, no harm to introduce it.